pkg/importer/gcs-datasource.go

```diff
@@ -0,0 +1,228 @@
+package importer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/option"
+
+	"k8s.io/klog/v2"
+
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+const (
+	gcsFolderSep = "/"
+	gcsScheme    = "gs"
+)
+
+// Helper for unit-testing
+var newReaderFunc = getGcsObjectReader
+
+// GCSDataSource is the struct containing the information needed to import from a GCS data source.
+// Sequence of phases:
+// 1. Info -> Transfer
+// 2. Transfer -> Convert
+type GCSDataSource struct {
+	// GCS end point
+	ep *url.URL
+	// Key File
+	keyFile string
+	// Reader
+	gcsReader io.ReadCloser
+	// stack of readers
+	readers *FormatReaders
+	// The image file in scratch space.
+	url *url.URL
+}
+
+// NewGCSDataSource creates a new instance of the GCSDataSource
+func NewGCSDataSource(endpoint, keyFile string) (*GCSDataSource, error) {
+	klog.V(3).Infoln("GCS Importer: New Data Source")
+	ep, err := ParseEndpoint(endpoint)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("GCS Importer: unable to parse endpoint %q", endpoint))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Getting Context
+	ctx := context.Background()
+
+	// Using gs:// endpoint
+	if strings.HasPrefix(endpoint, "gs") {
```

Comment:
its probably better to `url.Parse` and check the scheme like I suggested in the validator. Plus now you can set the url in the struct to it.

Reply:
Second there are essentially 3 things you are interested in both scenarios:
1. bucket
2. object
3. client
And the client can have an option indicating the withHost in the other case. I think we can simplify this if else into something like this:
```go
var bucket, object, host string
var options []option.ClientOption
// We assume URL will parse.
url, err := url.Parse(endpoint)
if err != nil {
  klog.Errorf("GCS Importer: error parsing url")
  return nil, err
}
if url.Scheme == "gs" {
  bucket, object = extractGcsBucketAndObject(endpoint)
} else if url.Scheme == "https" {
  bucket, object, host = extractGcsBucketObjectAndHost(endpoint)
  options = append(options, option.WithEndpoint(host))
}
client, err := getGcsClientGs(ctx, keyFile, options...)
if err != nil {
  klog.Errorf("GCS Importer: Error creating GS Client")
  return nil, err
}
defer client.Close() <--- we sure need this here? We have a close() elsewhere that should close the client
gcsReader, err := newReaderFunc(ctx, client, bucket, object)
if err != nil {
  klog.Errorf("GCS Importer: Error creating Reader")
  return nil, err
}
return &GCSDataSource{
  ep:        ep,
  keyFile:   keyFile,
  gcsReader: gcsReader,
  url: url,
}, nil
```

Reply:
fixed